File: dartcode/DataStructures/BinarySearchTree.dart
import 'dart:core';

class BstNode<T> {
  T key;
  BstNode<T> left;
  BstNode<T> right;

  BstNode(T value) {
    key = value;
    left = null;
    right = null;
  }

  Map<String, dynamic> toJson() => _nodeToJson(this);

  Map<String, dynamic> _nodeToJson(BstNode instance) => <String, dynamic>{
        'key': instance.key,
        'left': instance.left?.toJson(),
        'right': instance.right?.toJson()
      };

// adicionar verificacao para saber se valores sao de tipos comparaveis
  bool equals(T key) {
    return this.key == key;
  }

  bool lessThan(T key) {
    return (this.key as double) < (key as double);
  }

  bool biggerThan(T key) {
    return (this.key as double) > (key as double);
  }
}

class BinarySearchTree<T> {
  BstNode<T> root;

  BinarySearchTree() {
    this.root = null;
  }

  void insert(T key) {
    if (this.root == null) {
      this.root = new BstNode<T>(key);
    } else {
      this.insertNode(this.root, key);
    }
  }

  void insertNode(BstNode<T> node, T key) {
    if (node.biggerThan(key)) {
      if (node.left == null) {
        node.left = new BstNode<T>(key);
      } else {
        this.insertNode(node.left, key);
      }
    } else {
      if (node.right == null) {
        node.right = new BstNode<T>(key);
      } else {
        this.insertNode(node.right, key);
      }
    }
  }

  List<T> inOrderTraverse() {
    var listNode = new List<T>();
    this.inOrderTraverseNode(this.root, listNode);
    return listNode;
  }

  void inOrderTraverseNode(BstNode<T> node, List<T> listNode) {
    if (node != null) {
      this.inOrderTraverseNode(node.left, listNode);
      listNode.add(node.key);
      this.inOrderTraverseNode(node.right, listNode);
    }
  }

  List<T> preOrderTraverse() {
    var listNode = new List<T>();
    this.preOrderTraverseNode(this.root, listNode);
    return listNode;
  }

  void preOrderTraverseNode(BstNode<T> node, List<T> listNode) {
    if (node != null) {
      listNode.add(node.key);
      this.preOrderTraverseNode(node.left, listNode);
      this.preOrderTraverseNode(node.right, listNode);
    }
  }

  List<T> postOrderTraverse() {
    var listNode = new List<T>();
    this.postOrderTraverseNode(this.root, listNode);
    return listNode;
  }

  void postOrderTraverseNode(BstNode<T> node, List<T> listNode) {
    if (node != null) {
      this.postOrderTraverseNode(node.left, listNode);
      this.postOrderTraverseNode(node.right, listNode);
      listNode.add(node.key);
    }
  }

  BstNode<T> min() {
    return this.minNode(this.root);
  }

  BstNode<T> minNode(BstNode<T> node) {
    var current = node;
    while (current != null && current.left != null) {
      current = current.left;
    }
    return current;
  }

  BstNode<T> max() {
    return maxNode(this.root);
  }

  BstNode<T> maxNode(BstNode<T> node) {
    var current = node;
    while (current != null && current.right != null) {
      current = current.right;
    }
    return current;
  }

  bool search(T key) {
    return searchNode(this.root, key);
  }

  bool searchNode(BstNode<T> node, T key) {
    if (node == null) {
      return false;
    }
    if (node.biggerThan(key)) {
      return searchNode(node.left, key);
    } else if (node.lessThan(key)) {
      return searchNode(node.right, key);
    } else {
      return true;
    }
  }

  void remove(T key) {
    this.root = this.removeNode(this.root, key);
  }

  BstNode<T> removeNode(BstNode<T> node, T key) {
    if (node == null) {
      return null;
    }
    if (node.biggerThan(key)) {
      node.left = this.removeNode(node.left, key);
      return node;
    } else if (node.lessThan(key)) {
      node.right = this.removeNode(node.right, key);
      return node;
    } else {
      if (node.left == null && node.right == null) {
        node = null;
        return node;
      }
      if (node.left == null) {
        node = node.right;
        return node;
      } else if (node.right == null) {
        node = node.left;
        return node;
      }

      var aux = this.minNode(node.right);
      node.key = aux.key;
      node.right = this.removeNode(node.right, aux.key);
      return node;
    }
  }

   void printBST() {
    print(this.root?.toJson());
  }
}

void main(List<String> args) { 

BinarySearchTree<double> binarySearchTree = new BinarySearchTree<double>();
binarySearchTree.insert(2.3);
binarySearchTree.insert(3.1);
binarySearchTree.insert(1.9);
binarySearchTree.insert(2.3);
binarySearchTree.insert(0.7);
binarySearchTree.insert(4.6);
print("Binary Search Tree...\n");
binarySearchTree.printBST();

print("in order traverse...\n");
for (var x in binarySearchTree.inOrderTraverse()) {  
    print(x);
}

print("pre order traverse...\n");

for (var x in binarySearchTree.preOrderTraverse()) {  
    print(x);
}

print("post order traverse...\n");
for (var x in binarySearchTree.postOrderTraverse()) {  
    print(x);
}

print("min...\n");
print(binarySearchTree.min().toJson());
print("max...\n");
print(binarySearchTree.max().toJson());
print("searching 0.7...\n");
print(binarySearchTree.search(0.7));
print("searching 10...\n");
print(binarySearchTree.search(10));
print("searching 2.3...\n");
print(binarySearchTree.search(2.3));
print("searching 0.4...\n");
print(binarySearchTree.search(0.4));
print("searching -1...\n");
print(binarySearchTree.search(-1));
print("bst...\n");
binarySearchTree.printBST();
print("removing 0.7...\n");
binarySearchTree.remove(0.7);
print("bst...\n");
binarySearchTree.printBST();
print("removing 3.1...\n");
binarySearchTree.remove(3.1);
print("bst...\n");
binarySearchTree.printBST();
}

File: tscode/src/DataStructures/binary-search-tree.ts
class BSTNode<T> {
    key: T;
    left: BSTNode<T> | null;
    right: BSTNode<T> | null;
    constructor(key: T) {
        this.key = key;
        this.left = null;
        this.right = null;
    }
    public toString(): string {
        return JSON.stringify(this);
    }

    public equals(key: T): boolean {
        return this.key == key;
    }
    public lessThan(key: T): boolean {
        return this.key < key;
    }
    public biggerThan(key: T): boolean {
        return this.key > key;
    }
}

export class BinarySearchTree<T>{
    root: BSTNode<T> | null;

    constructor() {
        this.root = null;
    }

    insert(key: T): void {
        if (this.root == null) {
            this.root = new BSTNode(key);
        }
        else {
            this.insertNode(this.root, key);
        }
    }
    insertNode(node: BSTNode<T>, key: T) {
        if (node.biggerThan(key) || node.equals(key)) {
            if (node.left == null) {
                node.left = new BSTNode(key);
            }
            else {
                this.insertNode(node.left, key);
            }
        }
        else {
            if (node.right == null) {
                node.right = new BSTNode(key);
            }
            else {
                this.insertNode(node.right, key);
            }
        }
    }

    inOrderTraverse(): Array<T> {
        let listNode = new Array<T>();
        this.inOrderTraverseNode(this.root, listNode);
        return listNode;
    }
    inOrderTraverseNode(node: BSTNode<T> | null, listNode: T[]) {
        if (node != null) {
            this.inOrderTraverseNode(node.left, listNode);
            listNode.push(node.key);
            this.inOrderTraverseNode(node.right, listNode);
        }
    }
    preOrderTraverse(): Array<T> {
        let listNode = new Array<T>();
        this.inOrderTraverseNode(this.root, listNode);
        return listNode;
    }
    preOrderTraverseNode(node: BSTNode<T> | null, listNode: T[]) {
        if (node != null) {
            listNode.push(node.key);
            this.inOrderTraverseNode(node.left, listNode);
            this.inOrderTraverseNode(node.right, listNode);
        }
    }
    postOrderTraverse(): Array<T> {
        let listNode = new Array<T>();
        this.inOrderTraverseNode(this.root, listNode);
        return listNode;
    }
    postOrderTraverseNode(node: BSTNode<T> | null, listNode: T[]) {
        if (node != null) {
            this.inOrderTraverseNode(node.left, listNode);
            this.inOrderTraverseNode(node.right, listNode);
            listNode.push(node.key);
        }
    }

    min(): BSTNode<T> {
        return this.minNode(this.root!);
    }

    private minNode(node: BSTNode<T>): BSTNode<T> {
        let current = node;
        while (current != null && current.left != null) {
            current = current.left;
        }
        return current;
    }

    max(): BSTNode<T> {
        return this.maxNode(this.root!);
    }

    private maxNode(node: BSTNode<T>): BSTNode<T> {
        var current = node;
        while (current != null && current.right != null) {
            current = current.right;
        }
        return current;
    }

    public search(key: T): boolean {
        return this.searchNode(this.root!, key);
    }

    private searchNode(node: BSTNode<T>, key: T): boolean {
        if (node == null) {
            return false;
        }
        if (node.biggerThan(key)) {
            return this.searchNode(node.left!, key);
        }
        else if (node.lessThan(key)) {
            return this.searchNode(node.right!, key);
        }
        else {
            return true;
        }
    }

    public remove(key: T): void {
        this.root = this.removeNode(this.root!, key);
    }

    private removeNode(node: BSTNode<T> | null, key: T): BSTNode<T> | null {
        if (node == null) {
            return null;
        }
        if (node.biggerThan(key)) {
            node.left = this.removeNode(node.left!, key);
            return node;
        }
        else if (node.lessThan(key)) {
            node.right = this.removeNode(node.right!, key);
            return node;
        }
        else {
            if (node.left == null && node.right == null) {
                node = null;
                return node;
            }
            if (node.left == null) {
                node = node.right;
                return node;
            }
            else if (node.right == null) {
                node = node.left;
                return node;
            }

            var aux = this.minNode(node.right);
            node.key = aux.key;
            node.right = this.removeNode(node.right, aux.key);
            return node;
        }
    }

    public printBST(): string {
        return this.root ? this.root.toString() : '';
    }
}

let binarySearchTree: BinarySearchTree<number> = new BinarySearchTree<number>();
binarySearchTree.insert(2.3);
binarySearchTree.insert(3.1);
binarySearchTree.insert(1.9);
binarySearchTree.insert(2.3);
binarySearchTree.insert(0.7);
binarySearchTree.insert(4.6);
console.log("Binary Search Tree...\n");
console.log(binarySearchTree.printBST());

console.log("in order traverse...\n");
binarySearchTree.inOrderTraverse().forEach(x => {
    console.log(x);
});

console.log("pre order traverse...\n");
binarySearchTree.preOrderTraverse().forEach(x =>{
    console.log(x);
});

console.log("post order traverse...\n");
binarySearchTree.postOrderTraverse().forEach(x =>{
    console.log(x);
});

console.log("min...\n");
console.log(binarySearchTree.min());
console.log("max...\n");
console.log(binarySearchTree.max());
console.log("searching 0.7...\n");
console.log(binarySearchTree.search(0.7));
console.log("searching 10...\n");
console.log(binarySearchTree.search(10));
console.log("searching 2.3...\n");
console.log(binarySearchTree.search(2.3));
console.log("searching 0.4...\n");
console.log(binarySearchTree.search(0.4));
console.log("searching -1...\n");
console.log(binarySearchTree.search(-1));
console.log("bst...\n");
console.log(binarySearchTree.printBST());
console.log("removing 0.7...\n");
binarySearchTree.remove(0.7);
console.log("bst...\n");
console.log(binarySearchTree.printBST());
console.log("removing 3.1...\n");
binarySearchTree.remove(3.1);
console.log("bst...\n");
console.log(binarySearchTree.printBST());